crates/iota-metrics/src/hardware_metrics.rs

```diff
@@ -0,0 +1,280 @@
+                Self::label("cpu_vendor_id", &self.cpu_vendor_id),
+                Self::label(
+                    "cpu_core_count",
+                    self.cpu_core_count.map_or_else(
+                        || "cpu_core_count_unavailable".to_string(),
+                        |c| c.to_string(),
+                    ),
+                ),
+                Self::label("cpu_arch", System::cpu_arch()),
+            ]
+            .into()
+        });
+        metric
+    }
+    fn cpu_metric_family(&self) -> MetricFamily {
+        let mut mf = MetricFamily::new();
+        mf.set_name("cpu_specs".to_owned());
+        mf.set_help("CPU specs (brand,vendor,cores)".to_owned());
+        mf.set_field_type(prometheus::proto::MetricType::COUNTER);
+        mf.set_metric(vec![self.cpu_metric()].into());
+        mf
+    }
+
+    fn system_metric(&self) -> Metric {
+        let mut metric = Metric::new();
+        metric.set_label({
+            vec![
+                Self::label("is_docker", self.is_docker.to_string()),
+                Self::label(
+                    "os_version",
+                    System::long_os_version()
+                        .unwrap_or_else(|| "os_version_unavailable".to_string()),
+                ),
+            ]
+            .into()
+        });
+        metric
+    }
+    fn system_metric_family(&self) -> MetricFamily {
+        let mut mf = MetricFamily::new();
+        mf.set_name("system_info".to_owned());
+        mf.set_help("System info (OS, version, is_docker, ...)".to_owned());
+        mf.set_field_type(prometheus::proto::MetricType::COUNTER);
+        mf.set_metric(vec![self.system_metric()].into());
+        mf
+    }
+}
+
+impl Collector for HardwareSpecs {
+    fn desc(&self) -> Vec<&Desc> {
+        let mut descs = Vec::new();
+        descs.extend(self.memory_collector.desc());
+        descs.extend(self.disk_collector.desc());
+        descs
+    }
+
+    fn collect(&self) -> Vec<prometheus::proto::MetricFamily> {
+        let mut mfs = Vec::new();
+        mfs.push(self.cpu_metric_family());
+        mfs.push(self.system_metric_family());
+        mfs.extend(self.memory_collector.collect());
+        mfs.extend(self.disk_collector.collect());
+        mfs
+    }
+}
+
+pub fn is_running_in_docker() -> bool {
+    // Check for .dockerenv file instead. This file exists in the debian:__-slim
+    // image we use at runtime.
+    Path::new("/.dockerenv").exists()
+}
+
+fn human_fmt_bytes(bytes: u64) -> String {
+    const UNITS: [&str; 6] = ["B", "KB", "MB", "GB", "TB", "PB"];
+
+    let mut value = bytes;
+    let mut unit_idx = 0;
+    // Shift right until we're below 1024^2 or reach end of units
+    while value >= (1024 * 1024) && unit_idx < UNITS.len() - 2 {
+        value >>= 10;
+        unit_idx += 1;
+    }
+    let value: f64 = value as f64 / 1024.0;
+    unit_idx += 1;
+
+    format!("{:.2} {}", value, UNITS[unit_idx])
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum HardwareMetricsErr {
```

Comment:
I personally would put this enum before the `register_hardware_metrics` func, but I let you decide.

Reply:
I've moved it now